src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/Bmi1.cs

```diff
@@ -35,7 +35,7 @@ public abstract class X64
             ///   BEXTR r64a, reg/m64, r64b
             /// This intrinisc is only available on 64-bit processes
             /// </summary>
-            public static ulong BitFieldExtract(ulong value, byte start, byte length) => BitFieldExtract(value, start, length);
+            public static ulong BitFieldExtract(ulong value, byte start, byte length) => BitFieldExtract(value, (ushort)(start | (length << 8)));
```

Comment:
Implement the 3-arg version in managed code, which significantly simplified containment codegen.

Reply:
How does codegen look for this intrinsic? Such a combination of small integer type might result in such messy codegen that the intrinsic would be practically useless for non-constant start and length.

Reply:
Yes, there are some inefficient codegen from small integer
```asm
IN0001: 000009 mov      rcx, 0x1F272F13200
IN0002: 000013 mov      rcx, gword ptr [rcx]
IN0003: 000016 call     TestFramework:BeginScenario(ref)
IN0004: 00001B mov      rcx, 0x7FF905B8EFB8
IN0005: 000025 call     CORINFO_HELP_NEWSFAST
IN0006: 00002A mov      rdi, rax
IN0007: 00002D mov      rcx, rdi
IN0008: 000030 call     ScalarTernOpTest__BitFieldExtractUInt64:.ctor():this
IN0009: 000035 mov      rdx, qword ptr [rdi+8]
IN000a: 000039 mov      r8, rdx
IN000b: 00003C movzx    r9, byte  ptr [rdi+16]
IN000c: 000041 mov      ecx, r9d
IN000d: 000044 movzx    rax, byte  ptr [rdi+17]
IN000e: 000048 mov      r10d, eax
IN000f: 00004B shl      r10d, 8
IN0010: 00004F or       ecx, r10d
IN0011: 000052 movzx    rcx, cx
IN0012: 000055 vbextr   r8, rcx, r8
IN0013: 00005A mov      qword ptr [V03+0x20 rsp+20H], r8
IN0014: 00005F mov      r8, 0x1F272F13200
IN0015: 000069 mov      r8, gword ptr [r8]
IN0016: 00006C mov      gword ptr [V03+0x28 rsp+28H], r8
IN0017: 000071 mov      r8d, r9d
IN0018: 000074 mov      r9d, eax
IN0019: 000077 mov      rcx, rsi
IN001a: 00007A call     ScalarTernOpTest__BitFieldExtractUInt64:ValidateResult(long,ubyte,ubyte,long,ref):this
IN001b: 00007F nop  
```

Reply:
Hmm, it doesn't look too bad. I'm not sure what's up with IN000a and IN000c but they look like artifacts of a rather fancy test setup, you won't get those if you just use local variables. The only real problem seems to be the useless IN0011, that's a consequence of the unfortunately typed control parameter. I think a properly implemented version of JIT's `optNarrowTree` should be able to get rid of that extra cast.

Reply:
> The only real problem seems to be the useless IN0011, 

Right, I think so. The problem is that JIT does not know the following consumer is a truncation and should eliminate the promotion. This issue is not specific to this intrinsic, we have detected it from other ones (e.g., `Sse41.Insert`).

Seems related to https://github.com/dotnet/coreclr/issues/13210

Note, JIT prints the disasm incorrectly, that actually should be
```asm
0:  41 c1 e2 08             shl    r10d,0x8
4:  41 0b ca                or     ecx,r10d
7:  0f b7 c9                movzx  ecx,cx
a:  c4 42 f0 f7 c0          bextr  r8,r8,rcx
```

Reply:
I will fix the new instruction print in this PR, logged `movzx` in  #21441.

Reply:
> I will fix the new instruction print in this PR,

Done.

Reply:
> The problem is that JIT does not know the following consumer is a truncation and should eliminate the promotion

I wouldn't expect the JIT to start understanding such intrinsic details. Instead, it should recognize that the expression produces a value where the upper 16 bits are already 0. Though of course, it would have been easier if this problem wasn't introduced in the first place by using the wrong parameter type.

> Seems related to #13210

Sort of, but that's more specific and the fix I have for that does not involve `optNarrowTree`.